### ISSUE TITLE
error: Monitor.o: No such file or directory

### ISSUE THREAD

**Issue by lancesimms** [neilbrown/mdadm#7]
I have tried compiling several different versions of mdadm (3.2.6, 3.3.1, 3.3.2).  They are all failing when trying to use the object file Monitor.o.  I get

error: Monitor.o: No such file or directory

I've tried renaming monitor.c to Monitor.c.  I also tried downloading the Monitor.c file that I found on Github and using that one to build Monitor.o, but I still get errors.  Can someone tell me what is going on before I start digging deeply to try and figure it out myself?

**Comment by neilbrown** [comment]
My guess is that you are using a case-insensitive filesystem to
hold the mdadm source code. e.g. VFAT.
Could that be the case?
Try doing the build in /tmp - that should be safe.

NeilBrown

**Comment by lancesimms** [comment]
Neil, thank you very much for the response.

I do not believe this is the issue.  I have tried compiling on two different versions of Linux.  Excerpts from the output of ‘mount’ and ‘uname-a’ are shown below

1) Debian system
/dev/sda1 on / type ext3 (rw,errors=remount-ro)
 Linux debian 2.6.32-5-686 #1 SMP Tue May 13 16:33:32 UTC 2014 i686 GNU/Linux

2) Bare kernel system
/dev/sda4 on / type ext3 (rw)
Linux aztec2 2.6.32-431.29.2.2chaos.ch5.2.x86_64 #1 SMP Thu Dec 18 17:47:12 PST 2014 x86_64 x86_64 x86_64 GNU/Linux

Neither of these worked.  I also tried compiling in /tmp and did not have success.

I think the error is due to the fact that the Makefile specifies two different “monitor” files.

OBJS =  mdadm.o config.o policy.o mdstat.o  ReadMe.o util.o maps.o lib.o \
        Manage.o Assemble.o Build.o \
        Create.o Detail.o Examine.o Grow.o Monitor.o dlink.o Kill.o Query.o \
        Incremental.o Dump.o \
        mdopen.o super0.o super1.o super-ddf.o super-intel.o bitmap.o \
        super-mbr.o super-gpt.o \
        restripe.o sysfs.o sha1.o mapfile.o crc32.o sg_io.o msg.o xmalloc.o \
        platform-intel.o probe_roms.o

MON_OBJS = mdmon.o monitor.o managemon.o util.o maps.o mdstat.o sysfs.o \
        policy.o lib.o \
        Kill.o sg_io.o dlink.o ReadMe.o super-intel.o \
        super-mbr.o super-gpt.o \
        super-ddf.o sha1.o crc32.o msg.o bitmap.o xmalloc.o \
        platform-intel.o probe_roms.o

If we are using a case-sensitive file system, there should be two files: Monitor.c and monitor.c, correct?  I tried copying monitor.c to Monitor.c in 3.3.2 and this did not help.  I got other error messages:

arm-cortexa8-linux-gnueabi-gcc -Wall -Werror -Wstrict-prototypes -Wextra -Wno-unused-parameter -ggdb -DSendmail=\""/usr/sbin/sendmail -t"\" -DCONFFILE=\"/etc/mdadm.conf\" -DCONFFILE2=\"/etc/mdadm/mdadm.conf\" -DMAP_DIR=\"/dev/.mdadm\" -DMAP_FILE=\"map\" -DMDMON_DIR=\"/dev/.mdadm\" -DFAILED_SLOTS_DIR=\"/dev/.mdadm/failed-slots\"   -DUSE_PTHREADS  -o mdadm mdadm.o config.o policy.o mdstat.o  ReadMe.o util.o maps.o lib.o Manage.o Assemble.o Build.o Create.o Detail.o Examine.o Grow.o Monitor.o dlink.o Kill.o Query.o Incremental.o Dump.o mdopen.o super0.o super1.o super-ddf.o super-intel.o bitmap.o super-mbr.o super-gpt.o restripe.o sysfs.o sha1.o mapfile.o crc32.o sg_io.o msg.o xmalloc.o platform-intel.o probe_roms.o
mdadm.o: In function `main':
/tmp/mdadm-3-3.2/mdadm.c:1469: undefined reference to `Monitor'
mdadm.o: In function `misc_scan':
/tmp/mdadm-3-3.2/mdadm.c:1715: undefined reference to `WaitClean'
mdadm.o: In function `misc_list':
/tmp/mdadm-3-3.2/mdadm.c:1800: undefined reference to `Wait'
/tmp/mdadm-3-3.2/mdadm.c:1802: undefined reference to `WaitClean'
Monitor.o: In function `signal_manager':
/tmp/mdadm-3-3.2/Monitor.c:166: undefined reference to `mgr_tid'
/tmp/mdadm-3-3.2/Monitor.c:166: undefined reference to `mgr_tid'
Monitor.o: In function `wait_and_act':
/tmp/mdadm-3-3.2/Monitor.c:570: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:570: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:576: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:576: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:590: undefined reference to `manager_ready'
/tmp/mdadm-3-3.2/Monitor.c:590: undefined reference to `manager_ready'
/tmp/mdadm-3-3.2/Monitor.c:612: undefined reference to `remove_pidfile'
/tmp/mdadm-3-3.2/Monitor.c:613: undefined reference to `exit_now'
/tmp/mdadm-3-3.2/Monitor.c:613: undefined reference to `exit_now'
/tmp/mdadm-3-3.2/Monitor.c:650: undefined reference to `update_queue'
/tmp/mdadm-3-3.2/Monitor.c:650: undefined reference to `update_queue'
/tmp/mdadm-3-3.2/Monitor.c:653: undefined reference to `update_queue'
/tmp/mdadm-3-3.2/Monitor.c:653: undefined reference to `update_queue'
/tmp/mdadm-3-3.2/Monitor.c:656: undefined reference to `update_queue'
Monitor.o:/tmp/mdadm-3-3.2/Monitor.c:656: more undefined references to `update_queue' follow
Monitor.o: In function `wait_and_act':
/tmp/mdadm-3-3.2/Monitor.c:656: undefined reference to `update_queue_handled'
/tmp/mdadm-3-3.2/Monitor.c:656: undefined reference to `update_queue_handled'
/tmp/mdadm-3-3.2/Monitor.c:657: undefined reference to `update_queue'
/tmp/mdadm-3-3.2/Monitor.c:657: undefined reference to `update_queue'
/tmp/mdadm-3-3.2/Monitor.c:666: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:666: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:673: undefined reference to `discard_this'
/tmp/mdadm-3-3.2/Monitor.c:673: undefined reference to `discard_this'
collect2: error: ld returned 1 exit status
make: *** [mdadm] Error 1

Do you have any idea what’s going on in light of all this?

The good news is that I was able to compile mdadm under version 2.6.4.

Lance


------------------

Lance Simms
Lawrence Livermore National Laboratory
7000 East Ave. MS L210, Livermore, CA 94550
Phone: 925-423-9650
Fax: 925-422-4141

**Comment by neilbrown** [comment]
monitor.c and Monitor.c are two very different files.  When you check out the git repository you should have both of them.  When you "make", they should separately create monitor.o and Monitor.o.
If that doesn't happen, then I cannot imagine what is wrong.  Some sort of case insensitivity thing, but clearly not in the filesystem.  Maybe 'make' is the problem?  What version do you have?

2.6.4 builds because monitor.c wasn't added until 3.0.

**Comment by lancesimms** [comment]
Neil, I am very sorry for wasting your time.  You were absolutely correct that it was a problem related to case-insensitivity.  I was decompressing the files on my Mac, and regardless of what utility I used, it would indeed only create monitor.c and not Monitor.c.  I tried using tar on my Linux machines instead and both files were extracted.

Again, I apologize for writing before I had exhausted all the possibilities myself.  I had always thought that the Mac HFS filesystem was case sensitive, so I didn’t even consider this as a possibility.  Feel free to call me a moron.

So far I’ve been very content with version 2.6.4.  I will upgrade to the latest version now.  Thank you for creating this excellent utility.


------------------

Lance Simms
Lawrence Livermore National Laboratory
7000 East Ave. MS L210, Livermore, CA 94550
Phone: 925-423-9650
Fax: 925-422-4141
